### ISSUE TITLE
Introduce New header match type to avoid Regex - Contains

### ISSUE THREAD

**Issue by shivanshu21** [envoyproxy/envoy#12590]
Title: Introduce New header match type to avoid Regex - Contains

Description:
For matching values in the header that might be somewhere in the middle of the header, the present option is to use Regex in the form `.*Search-Pattern.*`. This can cause catastrophic backtracking as described here:
https://github.com/envoyproxy/envoy/issues/7728

As a solution, in our build used in eBay Traffic engineering, we have introduced another header match type called contains which is based on absl::StrContains(). It works well for us.

The API is exactly the same as suffix match for headers:

```
  |   |   |  
  |   |   | // If specified, header match will be performed based on whether the header value contains
  |   |   | // the value specified in this option.
  |   |   | // Note: empty contains match is not allowed, please use present_match instead.
  |   |   | //
  |   |   | // Examples:
  |   |   | //
  |   |   | // * The contains *abc* matches the value *xyzabcdpqrs*, but not for *xyzbcdpqrs*.
  |   |   | string contains_match = 11 [(validate.rules).string.min_bytes = 1];
```

Please let me know if this is required/acceptable upstream, I can cherry-pick and raise a PR. Here is a snippet from our change to further illustrate:

```
   case HeaderMatchType::Contains:
        match = absl::StrContains(header->value().getStringView(), header_data.value_);
        break;
```

**Comment by mattklein123** [comment]
Yes this sounds fine. I would also add this to StringMatcher in parallel if possible. Ultimately we would like to converge these 2 matchers. Thank you!

**Comment by shivanshu21** [comment]
https://github.com/envoyproxy/envoy/pull/12623

I will add the StringMatcher changes to this PR as well, hence kept the PR as WIP.
